### ISSUE TITLE
Consider exposing node labels via downward api

### ISSUE THREAD

**Issue by simonswine** [kubernetes/kubernetes#40610]
**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.): downward api node labels

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): Feature request

I think the downward API should support exposing the node labels on which the pod it is running. While I understand that the Pod should not be coupled too much with the node it is scheduled to. But I think there are certain scenarios that require the pod  to have more information about it's host.

- ***failure-zones*** like rack names on premise or availability zones in cloud
- ***instance-types*** for adapting thread settings per instance type
- other labels like local ssd storage available 

This helps applications that are cloud-native to understand where they are located. While this could be achieved using API calls to the node object, I think it should be supported by the downward API. A node object would expose way more information that should not be available to any pod. In OpenShift's default configuration there is access to the Node objects prevented by default.

@mattbates @smarterclayton @bprashanth

**Comment by smarterclayton** [comment]
There *could* be security implications for some types of clusters (exposing information the admin doesn't want the pod to have, like the exact core architecture).  That said, for most deployments it doesn't seem completely unreasonable.  A per node setting might decide whether to expose it or to return an empty map.

**Comment by tpetracca** [comment]
This was previously explored in #25957

**Comment by dwilliams782** [comment]
Is this still being considered? I'm currently trying to find a sensible way to retrieve the zone a pod is scheduled in for rack awareness, as per the original use case.

**Comment by laurieodgers** [comment]
We're looking at doing this for a per-zone load balancer, and per-region load balancer on a single deployment. Our use case is that customers must be able to bind multiple times to multiple failure domains. This could be achieved with 3x deployments (one per zone) but it would be much nicer to only have to worry about a single deployment.

**Comment by discordianfish** [comment]
I think this should be brought up in a sig-node meeting, otherwise I doubt this will get attention and ultimately get closed off by @fejta-bot

**Comment by thockin** [comment]
I am in favor of something like this.  The real problem here is that most people seem to want this for topology information and that information is somewhat arbitrary.  Hostname/Zone/Region are codified, so they get the most attention, but there is some demand for other topology.  That leads me to assert that the common API is labels and not something special for topology info.

**Comment by bjhaid** [comment]
would folks be comfortable with kubelet having a configurable whitelist of node-labels that can be consumed by pods via the downward-api. I am constantly hitting snags that having this feature would have unblocked or prevented weird workarounds

**Comment by thockin** [comment]
Can we talk about concrete needs?  What snags are you hitting?

I have heard concrete need for topology info, which is stored as labels and is indistinguishable from other labels.  That (rightly) sets off alarms for security-conscious people.  Do we need to first-class topology?  I'm less confident in that.

The last thing we really want is another configuration knob that admins have to know about to be secure...

**Comment by erik-stephens** [comment]
Our use case is being able to inform elasticsearch pods about topology so that we can make sure shards get placed across failure domains (eg racks).  Currently, we have to contort ourselves through some awkward initContainers to get those details into elasticsearch configs.

**Comment by thockin** [comment]
ACK.  Another +1 for topology.  I am not sure the right answer.  Moving
topology to a distinct struct would make it easy to copy that info into
pods, but there's a lot of momentum around labels already, so we'd have to
do both for the indefinite future.

As far as I know, nobody is signed up to try to push that yet.

**Comment by bjhaid** [comment]
> Can we talk about concrete needs? What snags are you hitting?

Sorry for not providing an helpful case :) All my snags have all been topology related, app needs to know what region/az or datacenter it is deployed to.

> As far as I know, nobody is signed up to try to push that yet.

Where do I get to sign up for this :) I don't mind pushing it if I get guidance

**Comment by 2rs2ts** [comment]
My use case is for the EC2 instance ID (which is automatically added as a label by the cloud provider) to be added as an environment variable for a container. Since I lacked this feature I had to provide a workaround (granted my pod access to the AWS metadata FUSE we use at my company https://github.com/jszwedko/ec2-metadatafs) that I wasn't really keen on. Having pod labels available in the downward API just makes sense honestly, I don't know why we need to introduce brand new features for this :D

**Comment by thockin** [comment]
> Where do I get to sign up for this :) I don't mind pushing it if I get guidance

You sign up by saying "I'd like to tackle this". :)

As for guidance, the real problem is that I don't know what the right design is.  The way I would usually start such an effort is with a brainstorm doc - try to pin down the kernel of the requirements expressed here and then dream up 10 different ways we could solve it, no judgement on good/bad ideas.  Then write the most aggressive pros/cons you can for each.  Then we discuss.

e.g. 

1) we could add node labels as a downward API "fake" field path
1b) we could make new syntax for field paths that follow object ref pointers
2) we could literally copy node labels to pods
3) we could add a new node topology block and copy THAT into pods
4) we could add a link-local metadata API for use at run-time
5) we could add a new type of volume/envFrom source that is node-metadata
...

**Comment by bjhaid** [comment]
> You sign up by saying "I'd like to tackle this". :)

I'd like to tackle this :)

Thanks for all of this context, I'll start writing some doc and share as soon I have something

**Comment by intlabs** [comment]
Like @erik-stephens failure domain awareness would be a prime motivator for this: in our case Ceph.

**Comment by bjhaid** [comment]
For folks interested myself and @thockin had a brainstorm in this doc https://docs.google.com/document/d/1fH64mEfZH597luup-ZBfBNkiTVjfoFYGEa-G1G_TM6A/edit#, I wrote something that covered the last idea (which is Tim's preference): https://github.com/bjhaid/node-labels-to-files

**Comment by erik-stephens** [comment]
Thanks @bjhaid, this helps but for my use-case (elasticsearch), I will still need another initContainer in order to get those values into a form that elasticsearch will read (eg write updated config, inject environment variables).  Maybe I can use something like [envdir](https://cr.yp.to/daemontools/envdir.html).  Some soft goals I'm aiming for:

1. Minimal changes to vendor's image
1. Use default entrypoint & command
1. Minimal initContainers

Chiming in here in case others on this thread have ideas.  I was hoping there was a stock/std linux way to do what `envdir` does but I couldn't find.

**Comment by danielmotaleite** [comment]
we need this for 2 user cases:
- nodes zone info in pod labels, so service selector can deliver traffic inside the same zone instead of doing cross zone requests
- environment variable with node zone info, so cluster based apps can have the correct info. It can also be used to request the correct in-zone dns and add the correct labels to monitoring ( like /metrics) and logs 

While we can use a initcontainer to set this, it require giving pods `patch` permissions to affected pods and so opening a security hole to allow some hacker to change kubernetes setup


My proposal is to have a node "flag"/label/annotations that would enable label propagation and what labels are allowed to be propagated (say: `node-labels-propagation: false` (default) , `true` (all labels) , `'{"failure-domain.beta.kubernetes.io/zone","failure-domain.beta.kubernetes.io/region"}' `)

- This allow the admin to control the info node leak to pods.

Then in the pods, it would be required to request what info they want, with two annotation, one for labels and another for environment ( say: `pod-labels-from-nodes-labels: false, true , '{"...","..."}' ` for label copy and `pod-enviroment-from-nodes-labels: false, true, '{..}' ` for environment). If needed, another option for a volume with the info (`pod-volume-from-node-labels: ...`)

- This allow the pod admin to request only what is needed and ignore any other label that he doesn't need

**Comment by Bessonov** [comment]
I'm not sure, that `fejta-bot` is useful. Why every one subscriber must get this message, just because no body take care about issues and hope that they are automagically get closed by time?

/remove-lifecycle stale

**Comment by discordianfish** [comment]
@Bessonov tell that @fejta 
I've just stopped filling issues I don't have the resources to fix right away.

**Comment by hzxuzhonghu** [comment]
One similar use case from istio: to support locality load balancing, we need to watch all nodes of the cluster to acquire its topology labels such as`[topology.kubernetes.io/region, topology.kubernetes.io/zone]`. We just need these locality info, but we have to get all the nodes, which is a waste of lot of memory. It would be better if k8s provide a way to allow setting these topology info directly into pod.

**Comment by jpmcb** [comment]
+1 for this as well

We would like access to this information in order for a highly available data store service to know which `[topology.kubernetes.io/region, topology.kubernetes.io/zone]` the data store processes are running. Without this information, we could end up in a scenario where data is stored in only a single region or zone. Upon that region becoming unavailable, it's possible the data store service would be 0% available.

**Comment by tedyu** [comment]
```
node-labels-propagation: false (default) , true (all labels) ...
```
I think when node label is allowed to propagate, 'all labels' should be avoided since there may be information which admin doesn't want to reveal.
Listing specific labels to propagate would serve the use cases.

**Comment by Feder1co5oave** [comment]
It is a bit sad seeing this feature, with so many practical use cases listed above, left unanswered for so long.
As a fully owned, single tenant Kubernetes user, I take care of both nodes AND pods, and I've been needing this feature for a long time.
It it utterly pointless to set topology information on nodes if that information can not be propagated to pods to implement intelligent, topology-aware deployments. Besides, pod affinity/antiaffinity rules already rely on node labels, and that is not considered a security concern.

I have never considered hiding nodes' labels a security feature.

I am wondering if we can tackle the security concerns by adding a clause that the user creating (or the serviceaccount associated with) a pod that requests node's information via one of the methods proposed by @bjhaid in the linked google doc, needs to have the `get node` permission?

**Comment by s97712** [comment]
> ```
> node-labels-propagation: false (default) , true (all labels) ...
> ```
> 
> I think when node label is allowed to propagate, 'all labels' should be avoided since there may be information which admin doesn't want to reveal.
> Listing specific labels to propagate would serve the use cases.

I think a better solution is allow to set different value of configMap/secret  for different nodes

**Comment by anyasabo** [comment]
As another +1 for topology info. [Topology spread constraints](https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints/) let you easily spread pods across different regions/racks etc, but there's no (easy) way for the pod to know where it is actually scheduled, so the application can't make smart decisions based on topology. Even making only `topology.kubernetes.io` labels available would be an improvement.

**Comment by iMartyn** [comment]
Another solid use-case is in a cluster-per-environment setup where a node label can be added (e.g. in gke or eks you could add a label `cluster-environment: staging` to all nodes in a cluster) and then that can be provided to workloads via the downward api to behave appropriately according to business logic.
This obviously doesn't fit with a `topology` structure, so I might propose that it is maybe slightly too narrow in scope.  Something perhaps to consider would be a format for "exposable" labels.  Perhaps something along the lines of `user/.*` being exposable.
Already regretting this comment just because of the notifications I'll get from fejta forever as it languishes.

**Comment by bozaro** [comment]
I want to have node zone name in envronment variable to use some latency-critical services strongly in the same zone (for example: `mongos`).

`mongos` is a separate service from mongodb, through which all requests in a sharded configuration pass (app <-> mongos <-> mongodb). Since these are database queries, each hop has a significant impact on overall performance.

In bare hardware we deploy and use mongos in the same host (`localhost`) for reduce traffic and latency.

In Kubernetes envrinoment we try to:
 - Reproduce localhost topology by DaemonSet deploy. This does not work well due to very different resource consumption from node to node and service to service.
 - Deploy mongos to same pod. This leads to too large RAM and CPU overhead.
 - Deploy separate mongos Service to every zone. This solution looks very ugly, as the application has to execute a lot of code to get the zone from the node's metadata. Moreover, this code is different in each language. If it was possible to add a zone to an environment variable, it would be much easier.

**Comment by steffeneichler** [comment]
Hi,

we need to propagate the node labels
failure-domain.beta.kubernetes.io/region
failure-domain.beta.kubernetes.io/zone
topology.kubernetes.io/region
topology.kubernetes.io/zone
to our pods because we need the information to distribute elasticsearch shards correctly and more safely.

So it would be great if this possibility could be implemented.

**Comment by howardjohn** [comment]
Given EndpointSlice is removing `topology`, this becomes even more important, as the only source of region/zone is now the node.

**Comment by hzxuzhonghu** [comment]
> Given EndpointSlice is removing topology

Any links? @howardjohn

**Comment by howardjohn** [comment]
https://opensource.googleblog.com/2020/11/kubernetes-efficient-multi-zone.html, https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2030-endpointslice-subsetting, and https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/2086-service-internal-traffic-policy/README.md

After thinking about it a bit more I don't think it is directly related to this issue though (needing to know the topology of all endpoints vs your own node labels) but this would still help in many other cases

**Comment by Feder1co5oave** [comment]
@steffeneichler

This is what I had to do to implement rack-awareness for elasticsearch in kubernetes:

Add rack awareness in elasticsearch.yml:
```yaml
cluster.routing.allocation.awareness.attributes: rack
```

Label my nodes with their rack-information:
```bash
kubectl label node my-node topology.kubernetes.io/rack=L101B09
```

Extract node information via kubectl, and export them to elasticsearch via env variable:
```yaml
apiVersion: apps/v1
kind: StatefulSet
spec:
  serviceName: elasticsearch
  template:
    spec:
      serviceAccountName: elasticsearch

      initContainers:
      - name: get-rack
        image: kubectl
        command:
        - bash
        - -c
        - |
          set -ex
          RACK=$(kubectl get node $NODE_NAME -o jsonpath='{.metadata.labels.topology\.kubernetes\.io/rack}')
          echo "$RACK" > /tmp/es-rack
        env:
        - name: NODE_NAME
          valueFrom:
            fieldRef:
              fieldPath: spec.nodeName
        volumeMounts:
        - name: tmp
          mountPath: /tmp/

      containers:
      - name: elasticsearch
        image: elasticsearch
        command:
        - bash
        - -c
        - |
          set -ex
          env "node.attr.rack=$(</tmp/es-rack)" /usr/local/bin/docker-entrypoint.sh eswrapper
        env:
        - name: node.name
          valueFrom:
            fieldRef:
              fieldPath: metadata.name
        - name: node.attr.node
          valueFrom:
            fieldRef:
              fieldPath: spec.nodeName
        volumeMounts:
        - name: tmp
          mountPath: /tmp/
      volumes:
      - name: tmp
        emptyDir: {}
```

Aaaaand of course you need RBAC to allow the serviceaccount to extract node information:
```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: elasticsearch
  namespace: elastic
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: elasticsearch-get-nodes
  namespace: elastic
rules:
- apiGroups: [""]
  resources: [nodes]
  verbs: ["get"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: elasticsearch-get-nodes
  namespace: elastic
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: elasticsearch-get-nodes
subjects:
- kind: ServiceAccount
  name: elasticsearch
  namespace: elastic
```

Yes, it is this ugly. I actually could have done it all in the main container by creating an image with both elasticsearch and kubectl.

**Comment by steffeneichler** [comment]
Hi @Feder1co5oave,

I found this way too. But this is a very ugly way. The best way would be to propagate the node topoloy information to the pods like the nodename.

**Comment by natereid72** [comment]
is this ever going to be addressed? perhaps a label for nodes prefaced with an indicator for downwardapi. e.g. dapi-region, dapi-az, etc. This way all of the node labels aren't exposed, but ones the admin wants are.

**Comment by immanetize** [comment]
Wow, this has been going for a long time.  I would really like to have a clear picture of how my pods are balanced between availability zones.  I can't imagine I am the only one operating kubernetes clusters in multiple failover zones who would benefit from this perspective.  Can anyone explain to me at a high level how region/AZ labels on metrics can be used to compromise the cluster or it's workload?  I don't understand the objection here, there are clear benefits to the proposal.

**Comment by thockin** [comment]
The simple answer is that it will get implemented when someone steps up to design and implement it.  There are hundreds and hundreds of these issues - really decent ideas - and they all need someone to step up to drive it.

I know that's not what you wanted to hear, but this project has a finite number of (mostly volunteer!) contributors with finite hours per week.  If you have some time, why not write a KEP?  It's not hard, but it needs some follow-thru.

In the interim, you can always run `kubectl get ... | jq ...` in a container in your pod, and save the output to a file on an emptydir volume.

**Comment by immanetize** [comment]
Absolutely, I do understand how community contributions work. I've read through two issues here with somewhat unsubstantiated privacy concerns, my hope was to get those concerns overtly documented here to help such contributions move forward.

**Comment by s97712** [comment]
What about the solution https://github.com/kubernetes/kubernetes/issues/93999 ?

**Comment by steffeneichler** [comment]
@s97712: I'm not sure how #93999 could help here to fix the problem. One thing is the node label and #93999 is about nodeSecret or configmap.

**Comment by s97712** [comment]
> @s97712: I'm not sure how #93999 could help here to fix the problem. One thing is the node label and #93999 is about nodeSecret or configmap.

Although the names are different, they are all data in essence.

**Comment by steffeneichler** [comment]
> 
> 
> > @s97712: I'm not sure how #93999 could help here to fix the problem. One thing is the node label and #93999 is about nodeSecret or configmap.
> 
> Although the names are different, they are all data in essence.

Everything is data. A node label is an existing node property. The task here is to find a way to expose this existing information to pod in the same way as the node name. #93999 is a request for new kind of information.

**Comment by laplasz** [comment]
in my point of view these node labels should be available by default as env variable...

**Comment by steffeneichler** [comment]
> 
> 
> in my point of view these node labels should be available by default as env variable...

that would be great.

**Comment by antaloala** [comment]
What about implementing it as a new kind of volume (e.g. nodeProperties) that a pod spec could require?
kubelet would be in charge of representing the requested node properties (in the volumes entry definition included in the Pod spec) as files, "mounting" it in the pod's container's specified path; those requested node properties (i.e. labels) configured in kubelet as "restricted" would be exposed to the pod's container as an empty file.

I think this is a cleaner solution than using the downward API (as it was thought for exposing to a pod's container some pod's fields, but not fields from another API object).

**Comment by m00lecule** [comment]
HI, also seeking for this feature. It also might be useful for reducing latency. Imagine a situation when we host each kafka node and k8s worker node in separate AZ. Then with additional metadata pod should pick the closest destination to provide the better network performance. I believe its a bit complicated, but this feature would make it possible.

**Comment by intjonathan** [comment]
>Imagine a situation when we host each kafka node and k8s worker node in separate AZ. Then with additional metadata pod should pick the closest destination to get a better network performance.

We're in exactly this situation and it's not just performance, but cost. Cross-AZ Kafka traffic is serious money at scale. Right now we're relying on GET requests to the node metadata API, but that's going to get locked down for compliance. A pattern for exposing certain fields and not others would satisfy both needs nicely.

**Comment by wmgroot** [comment]
It reads to me that a PR that allowed cluster admins to selective choose which metadata is made available to the downward API would address the security concerns in https://github.com/kubernetes/kubernetes/pull/25957#issuecomment-228041766

**Comment by derekwaynecarr** [comment]
I think its reasonable to open a KEP for this proposal for a specific set of allowed node labels for usage in downward API.  Labels have been better normalized since this issue was originally opened and expanding labels for particular use cases could be useful.  Allowing all labels though may prove problematic for users that label nodes with security sensitive information that is not to be exposed to the workload.

**Comment by schaefec** [comment]
Having access to a selected 'allow-listed' set of node labels through the downward API would address my use-case as well as long as the following [(Well-Known)](https://kubernetes.io/docs/reference/labels-annotations-taints/) labels would be part of this set:

[topology.kubernetes.io/zone](https://kubernetes.io/docs/reference/labels-annotations-taints/#topologykubernetesiozone)
[topology.kubernetes.io/region](https://kubernetes.io/docs/reference/labels-annotations-taints/#topologykubernetesioregion)

I understand and agree that exposing all node labels is definitely not wanted due to the concerns mentioned by @derekwaynecarr

Has a KEP been created for this already?

**Comment by vassilvk** [comment]
Injecting topology awareness into pods can be accomplished with the [KubeMod](https://github.com/kubemod/kubemod) operator.

Here's a KubeMod ModRule which intercepts the event of any pod being scheduled on a node, then copies the node's `topology.kubernetes.io/region` and `topology.kubernetes.io/zone` labels into the pod's labels before the pod's containers are created.

```yaml
apiVersion: api.kubemod.io/v1beta1
kind: ModRule
metadata:
  name: inject-topology-labels
  namespace: kubemod-system
spec:
  type: Patch
  targetNamespaceRegex: ".*"
  admissionOperations:
    - UPDATE

  match:
    # Match pods...
    - select: '$.kind'
      matchValue: 'Pod'
    # ...which have access to the node's manifest through the synthetic ref injected by KubeMod.
    - select: '$.syntheticRefs.node.metadata.labels'

  patch:
    # Grab the node's region and zone and put them in the pod's corresponding labels.
    - op: add
      path: /metadata/labels/topology.kubernetes.io~1region
      value: '"{{ index .Target.syntheticRefs.node.metadata.labels "topology.kubernetes.io/region"}}"'
    - op: add
      path: /metadata/labels/topology.kubernetes.io~1zone
      value: '"{{ index .Target.syntheticRefs.node.metadata.labels "topology.kubernetes.io/zone"}}"'
``` 

Having KubeMod and the above ModRule deployed to your cluster, transparently stamps out all pods with their node's topology labels whenever they get scheduled on a node. This makes topology information available to the pod's containers through Kubernetes Downward API.

For example, assuming that you have the above ModRule, the following Deployment will push the region and availability zone its pods are deployed to, into their containers' environment variables `TOPOLOGY_REGION` and `TOPOLOGY_ZONE`:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-deployment
spec:
  template:
    metadata:
      annotations:
        # Tell KubeMod to monitor this deployment's pods for scheduling events.
        ref.kubemod.io/inject-node-ref: "true"
    spec:
      containers:
      - name: my-container
        image: my-container-image:v1.2.3
        env:
          - name: TOPOLOGY_REGION
            valueFrom:
              fieldRef:
                # The following label will be populated by ModRule inject-topology-labels.
                fieldPath: metadata.labels['topology.kubernetes.io/region']
          - name: TOPOLOGY_ZONE
            valueFrom:
              fieldRef:
                # The following label will be populated by ModRule inject-topology-labels.
                fieldPath: metadata.labels['topology.kubernetes.io/zone']
```

**Comment by abirsigron** [comment]
@vassilvk KubeMode does the trick for me!
But there was a little mistake in your deployment.
the ref.kubemod.io/inject-node-ref: "true" annotation should be in the pod spec and not the deployment spec.

**Comment by fruitbutt** [comment]
6 years now, and there is still no way to access topology labels from the downward API. Wild.

**Comment by thockin** [comment]
https://github.com/kubernetes/kubernetes/issues/40610#issuecomment-1231988005

KEPs are welcome!

**Comment by ArangoGutierrez** [comment]
cc @marquiz 
/cc

**Comment by alculquicondor** [comment]
Not really. The Pod could always query the kubernetes API to get Node labels. Reading a NodeFeatures CRD has the same problem. The point of the downward API is that the Pod doesn't have to do that: the information is readily available to the Pod as an environment variable or a file.

**Comment by ArangoGutierrez** [comment]
we could drop that CR per node, so it can be consumed by pods, preventing pods from having to access the node it self. WDYT?

**Comment by ArangoGutierrez** [comment]
I can think in something like the `client-go` https://github.com/kubernetes/client-go/blob/master/rest/config.go#L511 
but to retrieve the NFD `NodeFeature` struck from the node

**Comment by alculquicondor** [comment]
The idea is that a workload doesn't need to know about the Kubernetes API, Node or CRD. The information should be available via environment variables, so you don't have to change your application or add a dependency on a kubernetes client.

**Comment by thockin** [comment]
The problem with downward API across resources is that the pod's SA might
intentionally NOT be allowed to read nodes.

Why not do this as a sidecar and an emptyDir volume?  We now have a kubectl
container image - just run that as an initcontainer, save the info you want
to a file, done.

If your pod doesn't have permission, it will fail.

**Comment by frittentheke** [comment]
It's totally understood, that the info is available in the Kubernetes API and that one could use an init container with some  Kubernetes API client to fetch it. This issue is about providing the info, just like https://github.com/kubernetes/kubernetes/issues/40610#issuecomment-1532052078 said, without adding extra components / using API clients. To me the sole purpose of the downward API is that NOT have to use an API client for rather common details (such as the pods requests and limits) or `spec.nodeName` and having to fiddle with the right policy to allow it.

And in a distributed environment (running applications across multiple nodes and also failure domains) it's not far fetched to also expose this info to the workload to e.g. replicate it's data accordingly. 

The discussion should be more about which info should not be exposed or needs further filtering and how this could be achieved without exploding into complexity hell, beating the purpose of this "elegant" solution of using the existing downloadAPI.

**Comment by thockin** [comment]
It's important to understand that this project has hundreds of large
features in flight, and thousands of open bugs and PRs. While it may seem
like an easy, obvious feature, it requires thinking about, testing, code
review, docs, etc.

Every feature we add has a cost, both to our ability to in get some other
things done and in complexity of the system.

Features/bugs which have viable workarounds are generally less critical
than those without.

As always, if you think this is important, a KEP or even a doc is the right
place to start proposing a design and evaluating the risks.

As a compromise, one thing we could probably do safely is copy the topology
labels (topology.kubernetes.io/*) from node to pod (making sure the pod
does not already set them itself) at scheduling time. To do more than that
would require more analysis.

**Comment by thockin** [comment]
Even this simple mode needs some thought. If a user sets a topology label
(they shouldn't but we don't stop them) should we override it or leave it?

Do we copy all such labels or just the well-known ones?

Do we copy labels for pods that manually set nodeName (no scheduling)?

Doel we copy labels onto mirror pods?

Will this cause problems for systems which enumerate pod labels to derive
identity (e.g. Cilium)?

Do we need to update those if the labels change on the node?  Topology
should not change, but we don't prevent it...

Just off the top of my head :)

**Comment by abursavich** [comment]
Automagically modifying pod labels is a bit scary to me, although it has the added benefit of free topology-based pod querying... How about adding a `NodeFieldRef *ObjectFieldSelector` field to `EnvVarSource` and `DownwardAPIVolumeFile` that is validated against an explicit allow-list of fields (which could grow in the future)?

e.g.:

- `metadata.name`
- `metadata.labels['topology.kubernetes.io/region']`
- `metadata.labels['topology.kubernetes.io/zone']`

**Comment by 2rs2ts** [comment]
@thockin re: https://github.com/kubernetes/kubernetes/issues/40610#issuecomment-1534903777 I'm not sure what the issue is here, we'd just be exposing node labels via the downward API right? So who cares what they're set to? Just make them available. We already have a downward API for things like node name. These are only exposable as env vars or mounted files. So this doesn't change pod labels. https://kubernetes.io/docs/concepts/workloads/pods/downward-api/

I believe in the past week this conversation has gotten off track from the original request. I understand that every feature _is_ an extra complexity and extra maintenance burden, but in this case I believe that the turn in discussion has made this much more complex than necessary.

I believe the only thing discussed in the past few posts that we are going to need to worry about is what happens when someone updates a node label after a pod that is using the label via the downward API has already fetched it. However, I'm _pretty_ sure that there are things in the downward API that can change out without the pod having to be recreated, and that's why they're not allowed to be environment variables. So, wouldn't this fall under that case?

---

Anyway to everyone here who's suggested a KEP, I want to say this, with the caveat that I'm not someone very important, I just want to give my opinion on a phenomenon I've witnessed as a user and small-time contributor over the past 5 years; nothing I'm saying should be taken personally or is even written with a particular group of people in mind since I've seen this come from so many contributors that I can't keep track of them.

The KEP process is so overly bureaucratic and tedious that it's no wonder that people avoid engaging with it, especially for proposals that are pretty small in scope. KEPs make sense for big changes like adding or removing entire features, or breaking changes to existing features, or other efforts that would be wide-in-scope; however, asking people to make them for every little thing makes it seem–to a lot of people requesting smaller changes (like this one)–that KEPs exist primarily to prevent improvements from being made without sufficiently flagellating the people who dared to be so brazen as to ask for that improvement. I have noticed that a lot of issues I follow end up leading down the "make a KEP (talk to the hand)" path. Of course that is not their stated purpose but I am not alone in perceiving this as their de facto purpose, or at least, an intended side effect.

I sympathize with everyone's concerns for not wanting scope creep to end up making maintenance a nightmare, of course! I believe in this particular case we are overthinking ourselves into reasons why not to accept something as a valid feature request. Since it's possible to work around the lack of this feature, I agree it'd be a low priority feature and have only a moderate value add, so it's reasonable to say it should be left to a community contributor to implement when they feel like they have the time, but invoking the KEP process before they're allowed to try an implementation would be (IMO) bikeshedding, wasting not only their time but also the time of everyone reviewing the KEP.

But, that's just my 2 cents as a user and an extremely minor contributor who doesn't attend SIG meetings and doesn't know anyone here personally. Perhaps my opinion is born out of ignorance of the inner workings of the core kubernetes contributor community. Hopefully you can take it as just one perspective out of many.

**Comment by thockin** [comment]
> we'd just be exposing node labels via the downward API right

How confident are you that every cluster operator in the world is OK with this - that none of them have information that isn't meant for pods to see?

> We already have a downward API for things like node name

...which are are all present as fields on the Pod object.  Node labels are fundamentally different, and warrant some caution.

> KEPs exist primarily to prevent improvements from being made

Almost correct - KEPs exist primarily to prevent improvements from being made without thinking about some of the harder and less obvious issues that are easy to miss.  I know it feels like a lot of BS, but the project is HUGE.  Even I don't know where all the traps are any more.  Everything in the KEP questionairre is there because someone forgot to think about it and caused a problem.  KEPs are how we communicate the details to each other across time and space.

 > invoking the KEP process before they're allowed to try an implementation

You are free to try anything you like.  Go implement it, and come back with evidence that it was easy, that all the tests Just Work, and use that to inform your KEP.  100%.  Go forth and hack.

> Hopefully you can take it as just one perspective out of many.

Look, I hear you.  I myself have said "oh no, another KEP to write".  But the reality is that the potential blast-radius of doing something dumb is just SO high, and it's so easy to make a mistake.  Writing a KEP is not really that hard, but it does make you think about some things you might like to avoid thinking about until later.  Bolting tests and metrics and scale on at the end just doesn't work.

I encourage you to try it.  It's not so bad.

**Comment by kerthcet** [comment]
I'm interest with this issue, let me try to make it out, a KEP or a design doc as a startup.
/assign

**Comment by jimmyjones2** [comment]
Or could spec.nodeTopologyZone and spec.nodeTopologyRegion be added from the nodes topology.kubernetes.io well known labels? Then pods could consume using downward API like spec.nodeName today?

**Comment by xvdy** [comment]
@kerthcet we need this feature. Does KEP in process? I want join this feature development.

**Comment by xvdy** [comment]
@kerthcet Sorry to interrupt. How is the progress now?

**Comment by marquiz** [comment]
Just a wild idea: how about making this a Node Resource Interface (NRI) plugin, instead(?) That is, not make it part of Kubernetes but a runtime plugin that can be deployed by those who need it. (Note: NRI is quite new and isn't enabled by default in cri-o or containerd, yet).

Refs:
- https://github.com/containerd/nri
-https://github.com/containerd/containerd/blob/main/docs/NRI.md
- https://kccncna2022.sched.com/event/182JT
- https://github.com/containers/nri-plugins

**Comment by ArangoGutierrez** [comment]
> Just a wild idea: how about making this a Node Resource Interface (NRI) plugin, instead(?) That is, not make it part of Kubernetes but a runtime plugin that can be deployed by those who need it. (Note: NRI is quite new and isn't enabled by default in cri-o or containerd, yet).
> 
> Refs:
> 
> * https://github.com/containerd/nri
>   -https://github.com/containerd/containerd/blob/main/docs/NRI.md
> * https://kccncna2022.sched.com/event/182JT
> * https://github.com/containers/nri-plugins

If we agree on this approach as a community, I am more than happy to take/assign this myself and work to get it done (with help from @marquiz )

cc @simonswine @kerthcet

**Comment by alculquicondor** [comment]
Would that make the solution less portable? How easy is it to install an NRI plugin, specially in a cloud environment?

**Comment by thockin** [comment]
Anything that bypasses access-control should be a non-starter.  Doing it as NRI or doing it in Kubelet or doing it in apiserver are all identical in this aspect.

As said elsewhere, I think that copying topology labels from node->pod is totally reasonable, but has some corner cases to sort out.  Anything more than that needs to answer access control.

As a workaround: https://github.com/thockin/kubectl-sidecar

**Comment by docandrew** [comment]
@thockin @SergeyKanzhelev we're working on a proof-of-concept solution to this issue and will have a PR for it - what's the best path forward to starting a discussion for approval? We'll plan on coming to the SIG node meetings or can coord via the Slack channel.

**Comment by thockin** [comment]
You can post a summary of the design and a link to a KEP here, to start?

**Comment by SergeyKanzhelev** [comment]
> what's the best path forward to starting a discussion for approval? We'll plan on coming to the SIG node meetings or can coord via the Slack channel.

SIG Node meeting, slack, mailing list all the good ways to collect feedback. As @thockin said - design and KEP links would be a good starting point. Before going deep into implementation, we definitely need an agreement on overall design. Off the opt of my head, security was the concern. Another thing that is always a question with the mutable values is how end user will know which node value was used at a time of scheduling and what is expected behavior on container restart.

**Comment by docandrew** [comment]
The design we're prototyping mimics the existing `metadata.labels` structure for both environment variables and downwardAPI volumes, but using a new `metadata.node.labels` FieldPath and doing the lookups from the *v1.Node object (for environment variables) or VolumeHost.GetNodeLabels() (for volumes) so I'd expect the behavior to match the pod label lookups, as far as mutability w/ the pod seeing updates as they happen. We'll submit a `enhancements` issue and start working on a KEP - excited to work with you all on this.

**Comment by thockin** [comment]
`metadata.node.labels` as a field path is weird -- it's not a real field.

https://github.com/kubernetes/kubernetes/issues/40610#issuecomment-2077964178

> I think that copying topology labels from node->pod is totally reasonable, but has some corner cases to sort out. Anything more than that needs to answer access control.

IOW - I don't think you can just expose ALL node labels.

**Comment by docandrew** [comment]
I think we're open to alternate field paths if there are suggestions. I'm curious though with the concerns about access control - are there examples where node labels can be abused by pods or have contained otherwise sensitive information?

**Comment by cmwylie19** [comment]
> `metadata.node.labels` as a field path is weird -- it's not a real field.
> 
> [#40610 (comment)](https://github.com/kubernetes/kubernetes/issues/40610#issuecomment-2077964178)
> 
> > I think that copying topology labels from node->pod is totally reasonable, but has some corner cases to sort out. Anything more than that needs to answer access control.
> 
> IOW - I don't think you can just expose ALL node labels.

Just to clarify i think he meant, `node.metadata.labels` to make it succinct with the actual spec. cc: @docandrew

**Comment by cmwylie19** [comment]
> `metadata.node.labels` as a field path is weird -- it's not a real field.
> 
> [#40610 (comment)](https://github.com/kubernetes/kubernetes/issues/40610#issuecomment-2077964178)
> 
> > I think that copying topology labels from node->pod is totally reasonable, but has some corner cases to sort out. Anything more than that needs to answer access control.
> 
> IOW - I don't think you can just expose ALL node labels.

RE: `I don't think you can just expose ALL node labels.` What would you think an appropriate way to handle exposing just a subset be? Maybe only certain FieldPaths?

**Comment by thockin** [comment]
> I'm curious though with the concerns about access control - are there examples where node labels can be abused by pods or have contained otherwise sensitive information?

Maybe I am over-blowi9ng it, but there are deployments where the average user can't read nodes.  I don't think we have the context to say whether it's OK for us to unilaterally publish information that the user was otherwise restricted from viewing.  If, by chance, there's some sensitive information in there - a tenant ID, a cloud provider ID, etc - and we expose it, I don't think that's OK.  I don't like living in fear, but I think we need to be cautious that we don't inadvertently expose our users.

> What would you think an appropriate way to handle exposing just a subset be? Maybe only certain FieldPaths?

I think I suggested elsewhere that we could copy `topology.k8s.io/*` from Node to Pod (as long as we handle collisions properly and document that there are not mutable after the fact) either when scheduling or during Kubelet pod-start.  The advantag I see in that is that downward API and projected volumes and so forth ALREADY WORK for labels.

**Comment by cmwylie19** [comment]
> > I think I suggested elsewhere that we could copy topology.k8s.io/* from Node to Pod (as long as we handle collisions properly and document that there are not mutable after the fact) either when scheduling or during Kubelet pod-start. The advantag I see in that is that downward API and projected volumes and so forth ALREADY WORK for labels.
> 
> What would prevent users from cloning any* label into `topology.k8s.io/*` ? still feels to me that exposing node labels via downward API will push a lot of customers to re-think their labels usage

Helper functions that utilize the existing functions, mainly, [`IsLegalSubscriptionAndQualifiedName`](https://github.com/cmwylie19/kubernetes/blob/0b17feb3c4f57a8d6e097020af36f929933c5b29/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L55) and [`SanitizeNodeLabels`](https://github.com/cmwylie19/kubernetes/blob/0b17feb3c4f57a8d6e097020af36f929933c5b29/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L39C6-L39C24). We would programmatically restrict the user from using illegal labels by the same means that are currently in the code base. This is more or less what we would be thinking, there is[ no official PR yet](https://github.com/kubernetes/kubernetes/compare/master...cmwylie19:kubernetes:node-labels-downward-api) because we have a [KEP](https://github.com/kubernetes/enhancements/pull/4747) PR right now.

Does this reduce your concerns?

**Comment by cmwylie19** [comment]
> > I think I suggested elsewhere that we could copy topology.k8s.io/* from Node to Pod (as long as we handle collisions properly and document that there are not mutable after the fact) either when scheduling or during Kubelet pod-start. The advantag I see in that is that downward API and projected volumes and so forth ALREADY WORK for labels.
> 
> What would prevent users from cloning any* label into `topology.k8s.io/*` ? still feels to me that exposing node labels via downward API will push a lot of customers to re-think their labels usage

also, not sure what you mean by cloning but in the event that you meant updating labels on a given node/pod there is no way to guarantee that the pod will have the latest label without a restart. In the KEP we discuss that we want to make that clear.

**Comment by thockin** [comment]
> What would prevent users from cloning any* label into topology.k8s.io/* ? 

If you are allowed to write to node labels, then go nuts. My point is that the average pod Creator may or may not be able to read node labels, and we should not assume that.

**Comment by ivankovnatsky** [comment]
Sorry for another notification caused, but we were also using another service at my prev company: https://github.com/danfromtitan/envars-from-node-labels quite successfully.

**Comment by abirsigron** [comment]
We encountered a similar requirement to expose our node labels to our pods.

For those in need of an immediate solution:
I successfully collaborated with the [kyverno](https://kyverno.io/) team to implement a robust [solution](https://kyverno.io/blog/2024/02/19/assigning-node-metadata-to-pods/)

Moving forward, I hope that Kubernetes will enhance the Downward API to support this requirement directly, thereby eliminating the need for external solutions like the one we developed with Kyverno.

**Comment by alculquicondor** [comment]
Folks, if you have the expertise, please open a KEP. But note that the thread already has a lot of open questions that need to be answered in the KEP to move forward.

**Comment by andrewsykim** [comment]
We should probably keep this open until new feature is GA

**Comment by airhorns** [comment]
Not sure if this is the best place to post this, but the current approach as it exists in 1.33 doesn't work super great with common k8s platforms like GKE and EKS. Both of those automatically set the `topology.kubernetes.io/zone` style labels, where as this KEP only forwards along `topology.k8s.io/zone` style labels. The providers set `.kubernetes.` but the KEP specified setting `.k8s.`. https://github.com/kubernetes/kubernetes/pull/127092#issuecomment-2827259030 suggested that the well known attributes seem wrong, which I can't really speak to, but I can say that the docs website and the providers setting them automatically makes them good candidates for automatic forwarding in my opinion! Otherwise, administrators using those platforms will have to duplicate the labels automatically set by the providers into the new namespace to be able to forward them down to pods.

**Comment by frittentheke** [comment]
@airhorns see https://github.com/kubernetes/kubernetes/pull/127092#issuecomment-2898583878

@munnerz confirmed this is a mistake and will fix this in a PR.

**Comment by SergeyKanzhelev** [comment]
This issue is now a KEP: https://github.com/kubernetes/enhancements/issues/4742

If the KEP solves the scenario, we can close this tracking issue. Anybody asking for it - please provide your feedback

**Comment by mmanciop** [comment]
> @mmanciop without reading the more than 200s comments, it seems the description of the issue just mentions that "exposing node topology labels via downward API" ... what is the scope that is missing?

The title of the issue is "node **labels**". Topology labels are a small subset.

**Comment by SergeyKanzhelev** [comment]
part of it is addressed by https://github.com/kubernetes/enhancements/issues/4742

Please give feedback

**Comment by SergeyKanzhelev** [comment]
KEP is in beta, closing the issue. Please reopen for other scenarios

/close
